lib/interface_quda.cpp

```diff
@@ -3252,346 +3230,389 @@ void invertQuda(void *hp_x, void *hp_b, QudaInvertParam *param)
-    // preconditioned source and reconstruction of the full solution are
-    // taken care of by Dirac::prepare() and Dirac::reconstruct(),
-    // respectively.
-
-    if (pc_solution && !pc_solve) {
-      errorQuda("Preconditioned (PC) solution_type requires a PC solve_type");
+    // Load gauge field after pushing the split communicator so the comm buffers, etc are setup according to
+    // the split topology.
+    printfQuda("Split grid loading gauge field...\n");
+    if (!is_staggered) {
+      loadGaugeQuda(collected_gauge->Gauge_p(), gauge_param);
+    } else {
+      freeGaugeQuda();
+      loadFatLongGaugeQuda(param, gauge_param, collected_milc_fatlink_field->Gauge_p(),
+                           collected_milc_longlink_field->Gauge_p());
     }
+    printfQuda("Split grid loaded gauge field...\n");
 
-    if (!mat_solution && !pc_solution && pc_solve) {
-      errorQuda("Unpreconditioned MATDAG_MAT solution_type requires an unpreconditioned solve_type");
+    if (param->dslash_type == QUDA_CLOVER_WILSON_DSLASH || param->dslash_type == QUDA_TWISTED_CLOVER_DSLASH
+        || param->dslash_type == QUDA_CLOVER_HASENBUSCH_TWIST_DSLASH) {
+      printfQuda("Split grid loading clover field...\n");
+      if (collected_clover) {
+        loadCloverQuda(collected_clover->V(false), collected_clover->V(true), param);
+      } else {
+        loadCloverQuda(nullptr, nullptr, param);
+      }
+      printfQuda("Split grid loaded clover field...\n");
     }
 
-    if (!mat_solution && norm_error_solve) {
-      errorQuda("Normal-error solve requires Mat solution");
+    for (int n = 0; n < param->num_src_per_sub_partition; n++) {
+      op(_collect_x[n]->V(), _collect_b[n]->V(), param, args...);
     }
 
-    if (param->inv_type_precondition == QUDA_MG_INVERTER && (pc_solve || pc_solution || !direct_solve || !mat_solution))
-      errorQuda("Multigrid preconditioning only supported for direct non-red-black solve");
+    profileInvertMultiSrc.TPSTART(QUDA_PROFILE_TOTAL);
+    profileInvertMultiSrc.TPSTART(QUDA_PROFILE_EPILOGUE);
+    push_communicator(default_comm_key);
+    updateR();
 
-    if (mat_solution && !direct_solve && !norm_error_solve) { // prepare source: b' = A^dag b
-      for(int i=0; i < param->num_src; i++) {
-        cudaColorSpinorField tmp((in->Component(i)));
-        dirac.Mdag(in->Component(i), tmp);
-      }
-    } else if (!mat_solution && direct_solve) { // perform the first of two solves: A^dag y = b
-      DiracMdag m(dirac), mSloppy(diracSloppy), mPre(diracPre);
-      SolverParam solverParam(*param);
-      Solver *solve = Solver::create(solverParam, m, mSloppy, mPre, mPre, profileInvert);
-      solve->blocksolve(*out,*in);
-      for(int i=0; i < param->num_src; i++) {
-        blas::copy(in->Component(i), out->Component(i));
-      }
-      delete solve;
-      solverParam.updateInvertParam(*param);
+    for (int d = 0; d < CommKey::n_dim; d++) {
```

Comment:
Why do you need to modify the gaugeparam here? Just trying to understand and I might miss something.

Reply:
`gauge_param->X` and `gauge_param->ga_pad` are expanded earlier in this function so here they are shrunk.